test/testsyncmove.cpp

```diff
@@ -660,7 +659,57 @@ private slots:
         QVERIFY(fakeFolder.syncOnce());
         QCOMPARE(fakeFolder.currentLocalState(), expectedState);
         QCOMPARE(fakeFolder.currentRemoteState(), expectedState);
-        */
+
+        // Reverse on remote
+        fakeFolder.remoteModifier().rename("A/AllEmpty", "AllEmpty");
+        fakeFolder.remoteModifier().rename("C/Empty", "Empty");
+        fakeFolder.remoteModifier().rename("C", "AllEmpty/C");
+        fakeFolder.remoteModifier().rename("A", "Empty/A");
+        expectedState = fakeFolder.currentRemoteState();
+        QVERIFY(fakeFolder.syncOnce());
+        QCOMPARE(fakeFolder.currentLocalState(), expectedState);
+        QCOMPARE(fakeFolder.currentRemoteState(), expectedState);
+    }
+
+    void testDeepHierarchy_data() {
+        QTest::addColumn<bool>("local");
+        QTest::newRow("remote") << false;
+        QTest::newRow("local") << true;
+    }
+
+    void testDeepHierarchy() {
+        QFETCH(bool, local);
+        FakeFolder fakeFolder{ FileInfo::A12_B12_C12_S12() };
+        auto &modifier = local ? fakeFolder.localModifier() : fakeFolder.remoteModifier();
+
+        modifier.mkdir("FolA");
+        modifier.mkdir("FolA/FolB");
+        modifier.mkdir("FolA/FolB/FolC");
+        modifier.mkdir("FolA/FolB/FolC/FolD");
+        modifier.mkdir("FolA/FolB/FolC/FolD/FolE");
+        modifier.insert("FolA/FileA.txt");
+        modifier.insert("FolA/FolB/FileB.txt");
+        modifier.insert("FolA/FolB/FolC/FileC.txt");
+        modifier.insert("FolA/FolB/FolC/FolD/FileD.txt");
+        modifier.insert("FolA/FolB/FolC/FolD/FolE/FileE.txt");
+        QVERIFY(fakeFolder.syncOnce());
+
+        modifier.insert("FolA/FileA2.txt");
+        modifier.insert("FolA/FolB/FileB2.txt");
+        modifier.insert("FolA/FolB/FolC/FileC2.txt");
+        modifier.insert("FolA/FolB/FolC/FolD/FileD2.txt");
+        modifier.insert("FolA/FolB/FolC/FolD/FolE/FileE2.txt");
+        modifier.rename("FolA", "FolA_Renamed");
+        modifier.rename("FolA_Renamed/FolB", "FolB_Renamed");
+        modifier.rename("FolB_Renamed/FolC", "FolA");
+        modifier.rename("FolA/FolD", "FolA/FolD_Renamed");
+        modifier.mkdir("FolB_Renamed/New");
+        modifier.rename("FolA/FolD_Renamed/FolE", "FolB_Renamed/New/FolE");
+        auto expected = local ? fakeFolder.currentLocalState() : fakeFolder.currentRemoteState();
+        QVERIFY(fakeFolder.syncOnce());
+        QCOMPARE(fakeFolder.currentLocalState(), expected);
+        QCOMPARE(fakeFolder.currentRemoteState(), expected);
```

Comment:
Do we already have a test for interacting renames on different sides?
```
local: A -> B
remote: A/foo -> A/bar
```
and the other way around.

Reply:
That's right,  but that'd be another issue.

Reply:
Fixed now

Reply:
Great, looks good